File: Java/src/main/java/org/codecop/lcdnumbers/Digit.java
package org.codecop.lcdnumbers;

import java.util.List;
import java.util.Objects;

/**
 * Value Object of an LCD digit.
 */
public class Digit {

    private final List<Line> lines;

    public Digit(int digit, Patterns patterns) {
        this(patterns.of(digit));
    }

    private Digit(List<Line> lines) {
        Objects.requireNonNull(lines);
        this.lines = lines;
    }

    public List<Line> lines() {
        return lines;
    }

    public ScaledDigit scale(List<ScaledLine> scaledLines) {
        Objects.requireNonNull(scaledLines);

        int linesCount = lines.size();
        int scaledCount = scaledLines.size();
        if (scaledCount < linesCount) {
            throw new IllegalArgumentException("Scaled lines must be more than original ones: " + scaledCount + ">=" + linesCount);
        }

        return new ScaledDigit(scaledLines);
    }

    @Override
    public String toString() {
        // debug
        return String.join("\n", lines.toArray(new String[0])) + "\n";
    }
}

File: JavaScript/src/digit.js
exports = typeof window !== 'undefined' && window !== null ? window : global;

// Java class Digit

/**
 * Value object of an LCD digit.
 */
function Digit(digit, patternsOf) {
    'use strict';

    // TODO Objects.requireNonNull(digit);
    // TODO Objects.requireNonNull(patternsOf);

    var lines = patternsOf(digit);
    // TODO Objects.requireNonNull(lines);

    this.lines = function() {
        return lines;
    };

    this.scale = function(scaledLines) {
        // TODO Objects.requireNonNull(scaledLines);

        var linesCount = lines.length,
            scaledCount = scaledLines.length;
        if (scaledCount < linesCount) {
            throw new RangeError('Scaled lines must be more than original ones: ' + scaledCount + '>=' + linesCount);
        }

        function patternOfScaledLines() {
            return scaledLines;
        }

        return new Digit(digit, patternOfScaledLines);
    };

    this.toString = function() {
        return String(digit);
    };

}

exports.Digit = Digit;
